public_html/wp-content/plugins/wc-post-types/inc/utilities.php

```diff
@@ -0,0 +1,70 @@
+<?php
+/**
+ * Utilities and helper functions for WordCamp Post Types.
+ *
+ * @package WordCamp\Post_Types\Utilities
+ */
+
+namespace WordCamp\Post_Types\Utilities;
+use WP_Post;
+
+defined( 'WPINC' ) || die();
+
+/**
+ * Get the user's avatar or featured image given a Speaker or Organizer post.
+ *
+ * @param int|WP_Post $post Post ID or post object.
+ * @param int         $size Size for avatar.
+ * @param string      $alt  Alt text to use, defaults to post title. Can be empty string to indicate empty alt.
+ *
+ * @return string
+ */
+function get_avatar_or_image( $post, $size, $alt = false ) {
+	global $wcpt_plugin;
+	$post = get_post( $post );
+	if ( ! $post ) {
+		return '';
+	}
+	if ( false === $alt ) {
+		$alt = get_the_title( $post->ID );
+	}
+
+	$email       = get_post_meta( $post->ID, '_wcb_speaker_email', true );
+	$user_id     = get_post_meta( $post->ID, '_wcpt_user_id', true );
+	$id_or_email = $email ? $email : $user_id;
+
+	// Get the gravatar source, or the default if no user info is set.
+	if ( $id_or_email ) {
+		$src = get_avatar_url( $id_or_email, array( 'size' => $size ) );
+	} else {
+		$src = get_avatar_url(
+			0,
+			array(
+				'size' => $size,
+				'force_default' => true,
+			)
+		);
+	}
+
+	$avatar = sprintf( '<img src="%1$s" alt="%2$s" />', esc_url( $src ), $alt );
```

Comment:
```suggestion
	$avatar = sprintf( '<img src="%1$s" alt="%2$s" />', esc_attr( $src ), esc_attr( $alt ) );
```

Reply:
Good catch, thanks!

Reply:
Is `esc_attr()` better for the `src` in this context? IIRC `esc_url()` is does both of these:

* URL-encodes the value
* makes it safe to use an an attribute

...while `esc_attr()` only does the latter function.

Reply:
> makes it safe to use an an attribute

🤔 Does it? It doesn't appear that `esc_url` calls either of the sanitizing functions that `esc_attr` does.

I'm also not clear on when `esc_url` is actually necessary. The two use cases seem to be for if you are displaying the URL directly in HTML (not the case when it's in an attribute) and when you are going to store it in the database. It's definitely possible that I'm missing something and there's more to it than that, though...

Reply:
I haven't dug into it myself, but it seems like trustworthy folks have concluded `esc_attr( esc_url())` is unnecessary:

- https://developer.wordpress.com/themes/escaping/
- https://developer.wordpress.org/reference/functions/esc_attr/#comment-2459

I'm not sure if it's _harmful_ to do _both_, but I wouldn't _only_ use `esc_attr()` on a URL.

Reply:
👍 Ok, I'd trust those sources.